### PR TITLE
[RFR] Fix link publication

### PR DESCRIPTION
**Pull request by ThieryMichel**

### CODE REVIEW

File: src/api/controller/api/publish.js [hunk 1]
```diff
@@ -1,20 +1,40 @@
...
+};
+
 export const publishMiddleware = async (ctx) => {
     const count = await ctx.dataset.count({});
-    let handled = 0;
 
     const columns = await ctx.field.findAll();
 
-    const transformDocument = ctx.getDocumentTransformer({ env: 'node', dataset: ctx.dataset }, columns);
+    const uriIndex = columns.findIndex(col => col.name === 'uri');
```
Comment (djhi) [inline src/api/controller/api/publish.js]: Can't you simply use `columns.find(col => col.name === 'uri')` here and pass the column found to `getDocumentTransformer` ?

File: src/api/controller/api/publish.js [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
-    }
+    await tranformAllDocument(
+        count,
+        ctx.dataset.findLimitFromSkip,
+        ctx.uriDataset.insertMany,
+        addUri,
+    );
+
+    const transformDocument = ctx
+        .getDocumentTransformer({ env: 'node', dataset: ctx.dataset }, columns.splice(uriIndex));
```
Comment (djhi) [inline src/api/controller/api/publish.js]: use `filter` instead of `splice`

File: src/api/models/uriDataset.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export default (db) => {
+    const collection = db.collection('dataset');
```
Comment (djhi) [inline src/api/models/uriDataset.js]: Why another model if targeting the same collection ?

**Review by djhi — CHANGES_REQUESTED** [review]
Forgot to add comments as changes requests..

File: src/api/controller/api/publish.js [hunk 4]
```diff
@@ -1,25 +1,82 @@
...
+import route from 'koa-route';
+
 /* eslint no-await-in-loop: off */
 import getDocumentTransformer from '../../../common/getDocumentTransformer';
 
-export const publishMiddleware = async (ctx) => {
-    const count = await ctx.dataset.count({});
+const app = new Koa();
+
+export const tranformAllDocument = async (count, findLimitFromSkip, insertBatch, transformer) => {
```
Comment (djhi) [inline src/api/controller/api/publish.js]: s/tranformAllDocument/tranformAllDocuments/

File: src/api/controller/api/publish.js [hunk 5]
```diff
@@ -1,25 +1,82 @@
...
+    uri: doc.uri,
+});
+
+export const doPublish = async (ctx) => {
+    const count = await ctx.dataset.count({});
 
     const columns = await ctx.field.findAll();
 
-    const transformDocument = ctx.getDocumentTransformer({ env: 'node' }, columns);
+    const uriIndex = columns.findIndex(col => col.name === 'uri');
```
Comment (djhi) [inline src/api/controller/api/publish.js]: use `find`

File: src/api/controller/api/publish.js [hunk 6]
```diff
@@ -1,25 +1,82 @@
...
+        ctx.dataset.findLimitFromSkip,
+        ctx.uriDataset.insertBatch,
+        addUri,
+    );
+
+    const transformDocument = ctx
+        .getDocumentTransformer({
+            env: 'node',
+            dataset: ctx.uriDataset,
+        }, columns.filter((_, index) => index !== uriIndex));
```
Comment (djhi) [inline src/api/controller/api/publish.js]: filter is good but you could filter on the name directly

File: src/api/controller/api/publish.js [hunk 7]
```diff
@@ -1,25 +1,82 @@
...
+        ctx.uriDataset.findLimitFromSkip,
+        ctx.publishedDataset.insertBatch,
+        transformDocumentAndKeepUri,
+    );
 
     ctx.redirect('/api/publication');
 };
 
-export default async (ctx) => {
+export const tryPublish = async (ctx, next) => {
```
Comment (djhi) [inline src/api/controller/api/publish.js]: `tryPublish` ? I expected a transformation to occur here. Isn't it an initializer for the post middleware ?

**Review by djhi — CHANGES_REQUESTED** [review]

File: src/api/controller/api/publish.js (hunk 7 above)
Comment (ThieryMichel) [reply src/api/controller/api/publish.js]: it both initialize ctx, and catch publish error, maybe I should split it in two, and rename to preparePublish and handlePublishError.

**Review by djhi — APPROVED** [review]

**Pull request merged**